libp2p/Session.cpp

```diff
@@ -202,57 +174,52 @@ bool Session::interpret(RLP const& _r)
 		if (m_server->id() == id)
 		{
 			// Already connected.
-			clogS(NetWarn) << "Connected to ourself under a false pretext. We were told this peer was id" << m_info.id.abridged();
+			clogS(NetWarn) << "Connected to ourself under a false pretext. We were told this peer was id" << id.abridged();
 			disconnect(LocalIdentity);
 			return true;
 		}
 
-		if (m_node && m_node->id != id)
+		// if peer and connection have id, check for UnexpectedIdentity
+		if (!id)
 		{
-			if (m_force || m_node->idOrigin <= Origin::SelfThird)
-				// SECURITY: We're forcing through the new ID, despite having been told
-				clogS(NetWarn) << "Connected to node, but their ID has changed since last time. This could indicate a MitM attack. Allowing anyway...";
-			else
-			{
-				clogS(NetWarn) << "Connected to node, but their ID has changed since last time. This could indicate a MitM attack. Disconnecting.";
-				disconnect(UnexpectedIdentity);
-				return true;
-			}
-
-			if (m_server->havePeer(id))
-			{
-				m_node->dead = true;
-				disconnect(DuplicatePeer);
-				return true;
-			}
+			disconnect(NullIdentity);
+			return true;
 		}
-
-		if (m_server->havePeer(id))
+		else if (!m_peer->id)
 		{
-			// Already connected.
-			clogS(NetWarn) << "Already connected to a peer with id" << id.abridged();
-			disconnect(DuplicatePeer);
+			m_peer->id = id;
+			m_peer->address.port(listenPort);
+		}
+		else if (m_peer->id != id)
+		{
+			disconnect(UnexpectedIdentity);
 			return true;
 		}
 
-		if (!id)
+		if (m_server->havePeerSession(id))
 		{
-			disconnect(NullIdentity);
+			// Already connected.
+			clogS(NetWarn) << "Already connected to a peer with id" << id.abridged();
+			disconnect(DuplicatePeer);
 			return true;
 		}
-
-		m_node = m_server->noteNode(id, bi::tcp::endpoint(m_socket.remote_endpoint().address(), listenPort), Origin::Self, false, !m_node || m_node->id == id ? NodeId() : m_node->id);
-		if (m_node->isOffline())
-			m_node->lastConnected = chrono::system_clock::now();
-		m_knownNodes.extendAll(m_node->index);
-		m_knownNodes.unionWith(m_node->index);
+		
+		if (m_peer->isOffline())
+			m_peer->lastConnected = chrono::system_clock::now();
 
 		if (m_protocolVersion != m_server->protocolVersion())
 		{
 			disconnect(IncompatibleProtocol);
 			return true;
 		}
-		m_info = PeerInfo({id, clientVersion, m_socket.remote_endpoint().address().to_string(), listenPort, std::chrono::steady_clock::duration(), _r[3].toSet<CapDesc>(), (unsigned)m_socket.native_handle(), map<string, string>() });
+		
+		m_info.clientVersion = clientVersion;
+		m_info.host = m_socket.remote_endpoint().address().to_string();
+		m_info.port = listenPort;
+		m_info.lastPing = std::chrono::steady_clock::duration();
+		m_info.caps = _r[3].toSet<CapDesc>();
+		m_info.socket = (unsigned)m_socket.native_handle();
```

Comment:
Static assert does not hurt, even for code that is planned to change later (which often never happens).